File: src/ExtraUtilities/worlds/blocks/heat/HeatDriver.java
package ExtraUtilities.worlds.blocks.heat;

import arc.Core;
import arc.graphics.g2d.Draw;
import arc.graphics.g2d.Lines;
import arc.graphics.g2d.TextureRegion;
import arc.math.Angles;
import arc.math.Mathf;
import arc.math.geom.Point2;
import arc.struct.IntSet;
import arc.struct.Seq;
import arc.util.Eachable;
import arc.util.Nullable;
import arc.util.Time;
import arc.util.io.Reads;
import arc.util.io.Writes;
import mindustry.Vars;
import mindustry.entities.units.BuildPlan;
import mindustry.gen.Building;
import mindustry.graphics.Drawf;
import mindustry.graphics.Pal;
import mindustry.ui.Bar;
import mindustry.world.Block;
import mindustry.world.blocks.heat.HeatBlock;
import mindustry.world.blocks.heat.HeatConductor;
import mindustry.world.blocks.heat.HeatConsumer;
import mindustry.world.draw.DrawBlock;
import mindustry.world.draw.DrawDefault;
import mindustry.world.meta.*;

import java.util.Arrays;

import static mindustry.Vars.*;

public class HeatDriver extends Block {
    public int range = 240;
    public float lost = 0.15f;
    public float visualMaxHeat = 15f;
    public DrawBlock drawer = new DrawDefault();
    public boolean splitHeat = false;

    public HeatDriver(String name) {
        super(name);

        sync = true;
        envEnabled |= Env.space;
        configurable = true;
        hasPower = true;
        update = solid = rotate = true;
        rotateDraw = false;
        size = 3;

        config(Point2.class, (HeatDriverBuild tile, Point2 point) -> tile.link = Point2.pack(point.x + tile.tileX(), point.y + tile.tileY()));
        config(Integer.class, (HeatDriverBuild tile, Integer point) -> tile.link = point);
    }

    @Override
    public void setStats() {
        super.setStats();
        stats.add(Stat.shootRange, (float) range / tilesize, StatUnit.blocks);
    }

    @Override
    public void drawPlace(int x, int y, int rotation, boolean valid) {
        super.drawPlace(x, y, rotation, valid);
        Drawf.dashCircle(x * tilesize + this.offset, y * tilesize + this.offset, range, Pal.accent);
    }
    @Override
    public void setBars(){
        super.setBars();
        addBar("heat", (HeatDriverBuild entity) -> new Bar(() -> Core.bundle.format("bar.heatamount", (int)(entity.heat + 0.001f)), () -> Pal.lightOrange, () -> entity.heat / visualMaxHeat));
    }

    @Override
    public void load(){
        super.load();

        drawer.load(this);
    }

    @Override
    public void drawPlanRegion(BuildPlan plan, Eachable<BuildPlan> list){
        drawer.drawPlan(this, plan, list);
    }

    @Override
    public TextureRegion[] icons(){
        return drawer.finalIcons(this);
    }

    public class HeatDriverBuild extends Building implements HeatBlock, HeatConsumer {
        public float rotation = 90f;
        public float progress = 0f;
        public float resProgress = 0f;
        public int link = -1;
        public Seq<Building> owners = new Seq<>();

        public float heat = 0f;
        public float[] sideHeat = new float[4];
        public IntSet cameFrom = new IntSet();
        public long lastHeatUpdate = -1;

        @Override
        public void draw(){
            drawer.draw(this);
        }

        @Override
        public void drawLight(){
            super.drawLight();
            drawer.drawLight(this);
        }


        @Override
        public void updateTile() {
            if(owners.size == 0 && link == -1) heat = 0;

            checkOwner();

            Building linked = world.build(this.link);
            boolean hasLink = linkValid();

            if(hasLink) {
                HeatDriverBuild other = (HeatDriverBuild) linked;
                if(other.checkOneOwner(this)) other.owners.add(this);
                float toRotation = this.angleTo(other);
                rotation = Mathf.slerpDelta(rotation, toRotation, 0.02f * power.status);
                if(Angles.near(rotation, toRotation, 2)){
                    this.updateTransfer();
                    //other.owners.add(this);
                    other.updateTransfer();
                    progress = Mathf.slerpDelta(progress, 1, 0.02f * power.status);
                } else {
                    progress = Mathf.slerpDelta(progress, 0, 0.04f);
                }
            } else {
                progress = Mathf.slerpDelta(progress, 0, 0.04f);
            }
            float p = Math.min((heat/visualMaxHeat), 1);
            if(owners.size > 0 && p > 0){
                resProgress = Mathf.slerpDelta(resProgress, 1, 0.02f * p);
            } else {
                resProgress = Mathf.slerpDelta(resProgress, 0, 0.05f);
            }
        }

        public void updateTransfer(){
            if(owners.size > 0){
                float totalHeat = 0f;
                for(int i = 0; i < owners.size; i++){
                    Building o = owners.get(i);
                    //if(o instanceof HeatDriverBuild){
                        HeatDriverBuild owner = (HeatDriverBuild)o;
                        float dst = this.dst(owner)/range;
                        if(Angles.near(owner.rotation, owner.angleTo(this), 2f)) {
                            totalHeat += owner.heat * (1 - dst * lost);
                            totalHeat *= owner.power.status;
                        //}
                    }
                }
                heat = totalHeat;
            } else {
                updateHeat();
            }
        }

        public void updateHeat(){
            if(lastHeatUpdate == Vars.state.updateId) return;

            lastHeatUpdate = Vars.state.updateId;
            heat = calculateHeat(sideHeat, cameFrom);
        }

        @Override
        public float heatRequirement(){
            return linkValid() ? visualMaxHeat : Float.MAX_VALUE;
        }

        @Override
        public float warmup(){
            return heat;
        }

        @Override
        public float heat(){
            return (owners.size > 0 && link == -1)? heat : 0;
        }

        @Override
        public float heatFrac(){
            return (heat / visualMaxHeat) / (splitHeat ? 3f : 1);
        }

        @Override
        public float[] sideHeat(){
            return sideHeat;
        }

        @Override
        public void drawConfigure(){
            float sin = Mathf.absin(Time.time, 6f, 1f);

            Draw.color(Pal.accent);
            Lines.stroke(1f);
            Drawf.circles(x, y, (tile.block().size / 2f + 1) * tilesize + sin - 2f, Pal.accent);

            owners.each(owner -> {
                Drawf.circles(owner.x, owner.y, (tile.block().size / 2f + 1) * tilesize + sin - 2f, Pal.place);
                Drawf.arrow(owner.x, owner.y, x, y, size * tilesize + sin, 4f + sin, Pal.place);
            });


            if(linkValid()){
                Building target = world.build(link);
                Drawf.circles(target.x, target.y, (target.block().size / 2f + 1) * tilesize + sin - 2f, Pal.place);
                Drawf.arrow(x, y, target.x, target.y, size * tilesize + sin, 4f + sin);
            }

            Drawf.dashCircle(x, y, range, Pal.accent);
        }

        @Override
        public boolean onConfigureBuildTapped(Building other){
            if(this == other){
                if(link == -1) deselect();
                configure(-1);
                return false;
            }

            if(link == other.pos()){
                configure(-1);
                return false;
            }else if(other.block == block && other.dst(tile) <= range && other.team == team && checkOneOwner(other)){
                configure(other.pos());
                return false;
            }

            return true;
        }

        public void checkOwner(){
            for(int i = 0; i<owners.size; i++){
                int pos = owners.get(i).pos();
                Building build = world.build(pos);
                if (build instanceof HeatDriverBuild owner) {
                    if (owner.block != block || owner.link != this.pos()) owners.remove(i);
                } else {
                    owners.remove(i);
                }
            }
        }

        protected boolean checkOneOwner(Building other){
            if(owners.size == 0) return true;
            return !owners.contains(other);
        }

        @Override
        public Point2 config() {
            if(tile == null) return null;
            return Point2.unpack(link).sub(tile.x, tile.y);
        }

        public boolean linkValid(){
            if(link == -1) return false;
            Building other = world.build(link);
            if(other == null){
                link = -1;
                return false;
            }
            return other.block == block && other.team == team && within(other, range);
        }

        @Override
        public void write(Writes write) {
            super.write(write);
            write.i(link);
            write.f(rotation);
            write.f(progress);
        }

        @Override
        public void read(Reads read, byte revision){
            super.read(read, revision);
            link = read.i();
            rotation = read.f();
            progress = read.f();
        }
    }
}

File: assets/scripts/block/heat/heatDriver.js
const lib = require("blib");
const Drawer = require("drawer/DrawHeatDriver");

const range = 45 * 8;//÷8格
const arrowSpacing = 4, arrowOffset = 2, arrowPeriod = 0.4, arrowTimeScl = 6.2;

const heatDriver = extend(HeatConductor, "heat-driver", {
    drawPlace(x, y, rotation, valid) {
        this.super$drawPlace(x, y, rotation, valid);
        Drawf.dashCircle(x * Vars.tilesize + this.offset, y * Vars.tilesize + this.offset, range, Pal.accent);
    },
    setStats(){
        this.super$setStats();
        this.stats.add(Stat.shootRange, range, StatUnit.blocks);
    },
    drawPlanRegion(plan, list){
        this.super$drawPlanRegion(plan, list);
        Draw.rect(Core.atlas.find("btm-heat-driver-outline"), plan.drawx(), plan.drawy());
    },
    icons(){
        return [Core.atlas.find("btm-heat-driver"), Core.atlas.find("btm-heat-driver-outline")];
    },
});
heatDriver.size = 3;
heatDriver.configurable = true;
heatDriver.hasPower = true;
heatDriver.consumePower(4);
heatDriver.visualMaxHeat = 30;
heatDriver.drawer = new DrawMulti(new DrawDefault(), new DrawHeatOutput(), new DrawHeatInput("-heat"), Drawer.DrawHeatDriver(4, 2, 0.4, 6.2));
heatDriver.requirements = ItemStack.with(
    Items.tungsten, 150,
    Items.carbide, 50,
    Items.oxide, 125
);
heatDriver.buildVisibility = BuildVisibility.shown;
heatDriver.category = Category.crafting;
// heatDriver.config(Point2, lib.cons2((tile, point) => {
//     tile.setLink2(Point2.pack(point.x + tile.tileX(), point.y + tile.tileY()));
// }));
heatDriver.config(java.lang.Integer, lib.cons2((tile, int) => {
    tile.setOneLink(int);
}));


heatDriver.buildType = prov(() => {
    const block = heatDriver;
    var link = -1;
    var owner = new Seq();
    var rotation = 90;
    var progress = 0;
    var resProgress = 0;
    
    return new JavaAdapter(HeatConductor.HeatConductorBuild, {
        setOneLink(v){
            var int = new java.lang.Integer(v);
            link = int;
        },
        setLink2(v){
            link = v;
        },
        setOwner(build){
            owner.add(build);
        },
        removeOwner(build){
            owner.remove(build);
        },
        getOwner(){ return owner; },
        getLink(){
            return link;
        },
        getRotation(){
            return rotation;
        },
        getProgress(){
            return progress;
        },
        getResProgress(){
            return resProgress;
        },
        linkValid(){
            if(link == -1) return false;
            var other = Vars.world.build(link);
            if(other == null){
                link = -1;
                return false;
            }
            return other.block == block && other.team == this.team && this.within(other, range);
        },
        setHeat(v){
            this.heat = v;
        },
        updateTile(){
            this.checkOwner();
            if(owner.size == 0 && link == -1) this.heat = 0;
            var other = Vars.world.build(link);
            var hasLink = this.linkValid();
            if(hasLink){
                if(other.checkOneOwner(this)) other.setOwner(this);
                var torotation = this.angleTo(other);
                rotation = Mathf.slerpDelta(rotation, torotation, 0.02 * this.power.status);
                if(Angles.near(rotation, torotation, 2)){
                    this.updateTransfer();
                    other.updateTransfer();
                    progress = Mathf.slerpDelta(progress, 1, 0.02 * this.power.status);
                } else {
                    progress = Mathf.slerpDelta(progress, 0, 0.04);
                }
            } else {
                progress = Mathf.slerpDelta(progress, 0, 0.04);
            }
            var p = Math.min((this.heat/block.visualMaxHeat), 1);
            if(owner.size > 0 && p > 0){
                resProgress = Mathf.slerpDelta(resProgress, 1, 0.02 * p);
            } else {
                resProgress = Mathf.slerpDelta(resProgress, 0, 0.05);
            }
        },
        updateTransfer(){
            if(owner.size > 0){
                var heat = 0;
                for(var i = 0; i < owner.size; i++){
                    var owners = owner.get(i);
                    var dst = this.dst(owners)/range;
                    if(Angles.near(owners.getRotation(), owners.angleTo(this), 2)){
                        heat += owners.heat*(1 - dst * 0.15);
                        heat *= owners.power.status;
                    }
                }
                this.heat = heat;
            } else {
                this.super$updateTile();
            }
        },
        heat(){
            return owner.size > 0 && link == -1 ? this.heat : 0;
        },
        drawConfigure(){
            const sin = Mathf.absin(Time.time, 6, 1);
            Draw.color(Pal.accent);
            Lines.stroke(1);
            Drawf.circles(this.x, this.y, (block.size / 2 + 1) * Vars.tilesize + sin - 2, Pal.accent);
            if(this.linkValid()){
                var other = Vars.world.build(link);
                Drawf.circles(other.x, other.y, (block.size / 2 + 1) * Vars.tilesize + sin - 2, Pal.place);
                Drawf.arrow(this.x, this.y, other.x, other.y, block.size * Vars.tilesize + sin, 4 + sin);
            }
            for(var i = 0; owner.size > 0 && i < owner.size; i++){
                Drawf.circles(owner.get(i).x, owner.get(i).y, (block.size / 2 + 1) * Vars.tilesize + sin - 2, Pal.place);
                Drawf.arrow(owner.get(i).x, owner.get(i).y, this.x, this.y, block.size * Vars.tilesize + sin, 4 + sin, Pal.place);
            }
            Drawf.dashCircle(this.x, this.y, range, Pal.accent);
        },
        onConfigureBuildTapped(other){
            if(this == other){
                if(link == -1) this.deselect();
                this.configure(-1);
                return false;
            }

            if(link == other.pos()){
                other.removeOwner(this);
                this.configure(-1);
                return false;
            }else if(other.block == this.block && this.dst(other) <= range && other.team == this.team && this.checkOneOwner(other)){
                var lastLink = Vars.world.build(link);
                if(lastLink != null) lastLink.removeOwner(this);
                other.setOwner(this);
                this.configure(other.pos());
                return false;
            }

            return true;
        },
        configure(config){
            link = config;
        },
        checkOwner(){
            for(var i = 0; i < owner.size; i++){
                var pos = owner.get(i).pos();
                var build = Vars.world.build(pos);
                if(build == null || build.block != block || build.getLink() != this.pos()) owner.remove(i);
            }
        },
        checkOneOwner(other){
            if(owner.size == 0) return true;
            var i = 0;
            for(; i < owner.size; i++){
                var pos = owner.get(i).pos();
                if(Vars.world.build(pos) == other) break;
            }
            if(i == owner.size) return true;
            return false;
        },
        // config(){
        //     if(this.tile == null) return null;
        //     return new Packages.arc.math.geom.Point2.unpack(link).sub(this.tile.x, this.tile.y);
        // },
        write(write) {
            this.super$write(write);
            write.i(link);
            write.f(rotation);
            write.f(progress);
        },
        read(read, revision) {
            this.super$read(read, revision);
            link = read.i();
            rotation = read.f();
            progress = read.f();
        },
    }, heatDriver);
});

